mapeditor/validator.cpp

```diff
@@ -0,0 +1,162 @@
+#include "StdInc.h"
+#include "validator.h"
+#include "ui_validator.h"
+#include "../lib/mapObjects/MapObjects.h"
+#include "../lib/CHeroHandler.h"
+
+Validator::Validator(const CMap * map, QWidget *parent) :
+	QDialog(parent),
+	ui(new Ui::Validator)
+{
+	ui->setupUi(this);
+
+	show();
+	
+	setAttribute(Qt::WA_DeleteOnClose);
+
+	std::array<QString, 2> icons{"mapeditor/icons/mod-update.png", "mapeditor/icons/mod-delete.png"};
+
+	for(auto & issue : Validator::validate(map))
+	{
+		auto * item = new QListWidgetItem(QIcon(icons[issue.critical ? 1 : 0]), issue.message);
+		ui->listWidget->addItem(item);
+	}
+}
+
+Validator::~Validator()
+{
+	delete ui;
+}
+
+std::list<Validator::Issue> Validator::validate(const CMap * map)
+{
+	std::list<Validator::Issue> issues;
+	
+	if(!map)
+	{
+		issues.emplace_back("Map is not loaded", true);
+		return issues;
+	}
+	
+	try
+	{
+		//check player settings
+		int hplayers = 0;
+		int cplayers = 0;
+		std::map<int, int> amountOfCastles;
+		for(int i = 0; i < map->players.size(); ++i)
+		{
+			auto & p = map->players[i];
+			if(p.canAnyonePlay())
+				amountOfCastles[i] = 0;
+			if(p.canComputerPlay)
+				++cplayers;
+			if(p.canHumanPlay)
+				++hplayers;
+			if(p.allowedFactions.empty())
+				issues.emplace_back(QString("No factions allowed for player %1").arg(i), true);
+		}
+		if(hplayers + cplayers == 0)
+			issues.emplace_back("No players allowed to play this map", true);
+		if(hplayers + cplayers == 1)
```

Comment:
It is possible to have a map with only one player. Its goal can be to find grail for instance.

Reply:
Sure if your editor does not support special win conditions - it is correct that we need at least 2